Foxtrot/Driver/Rewriting/EmitAsyncClosure.cs

```diff
@@ -0,0 +1,969 @@
+
+            this.checkPostBody = new StatementList();
+            this.closureClass.Members.Add(this.checkPostMethod);
+
+            if (taskType.IsGeneric)
+            {
+                // Assign taskParameter to originalResultLocal because
+                // this field is used in a postcondition
+                checkPostBody.Add(new AssignmentStatement(this.originalResultLocal, checkMethodTaskParameter));
+            }
+        }
+
+        private void AddAsyncPost(List<Ensures> asyncPostconditions)
+        {
+            var origBody = new Block(this.checkPostBody);
+            origBody.HasLocals = true;
+
+            var newBodyBlock = new Block(new StatementList());
+            newBodyBlock.HasLocals = true;
+
+            var methodBody = new StatementList();
+            var methodBodyBlock = new Block(methodBody);
+            methodBodyBlock.HasLocals = true;
+
+            checkPostMethod.Body = methodBodyBlock;
+
+            methodBody.Add(newBodyBlock);
+            Block newExitBlock = new Block();
+            methodBody.Add(newExitBlock);
+
+            // Map closure locals to fields and initialize closure fields
+
+            foreach (Ensures e in asyncPostconditions)
+            {
+                if (e == null) continue;
+
+                this.Visit(e);
+
+                if (this.forwarder != null)
+                {
+                    this.forwarder.Visit(e);
+                }
+
+                ReplaceResult repResult = new ReplaceResult(
+                    this.checkPostMethod, this.originalResultLocal,
+                    this.rewriter.AssemblyBeingRewritten);
+
+                repResult.Visit(e);
+
+                if (repResult.ContractResultWasCapturedInStaticContext)
+                {
+                    this.contractResultCapturedInStaticContext.Add(e.Assertion.SourceContext);
+                }
+
+                // now need to initialize closure result fields
+                foreach (var target in repResult.NecessaryResultInitializationAsync(this.closureLocals))
+                {
+                    // note: target here 
+                    methodBody.Add(new AssignmentStatement(target, this.originalResultLocal));
+                }
+            }
+
+            // Emit normal postconditions
+
+            SourceContext? lastEnsuresSourceContext = null;
+
+            var ensuresChecks = new StatementList();
+
+            Method contractEnsuresMethod = this.rewriter.RuntimeContracts.EnsuresMethod;
+
+            foreach (Ensures e in GetTaskResultBasedEnsures(asyncPostconditions))
+            {
+                // TODO: Not sure that 'break' is enough! It seems that this is possible
+                // only when something is broken, because normal postconditions 
+                // are using Contract.Result<T>() and this is possible only for
+                // generic tasks.
+                if (IsVoidTask()) break; // something is wrong in the original contract
```

Comment:
This one was old behavior and I don't want to change existing behavior to much. I've faced a few issues when my obviously innocent change break existing code. The problem with fail-fast behavior for such a change that it could break existing customers.